util/globalconn/globalconn.go

```diff
@@ -46,4 +62,231 @@ const (
+
+type serverIDGetterFn func() uint64
+
+// Allocator allocates global connection IDs.
+type Allocator interface {
+	// NextID returns next connection ID.
+	NextID() uint64
+	// Release releases connection ID to allocator.
+	Release(connectionID uint64)
+	// GetReservedConnID returns reserved connection ID.
+	GetReservedConnID(reservedNo uint64) uint64
+}
+
+var (
+	_ Allocator = (*SimpleAllocator)(nil)
+	_ Allocator = (*GlobalAllocator)(nil)
 )
+
+// SimpleAllocator is a simple connection id allocator used when GlobalKill feature is disable.
+type SimpleAllocator struct {
+	pool AutoIncPool
+}
+
+// NewSimpleAllocator creates a new SimpleAllocator.
+func NewSimpleAllocator() *SimpleAllocator {
+	a := &SimpleAllocator{}
+	a.pool.Init(math.MaxUint64 - ReservedCount)
+	return a
+}
+
+// NextID implements ConnIDAllocator interface.
+func (a *SimpleAllocator) NextID() uint64 {
+	id, _ := a.pool.Get()
+	return id
+}
+
+// Release implements ConnIDAllocator interface.
+func (a *SimpleAllocator) Release(id uint64) {
+	a.pool.Put(id)
+}
+
+// GetReservedConnID implements ConnIDAllocator interface.
+func (*SimpleAllocator) GetReservedConnID(reservedNo uint64) uint64 {
+	if reservedNo >= ReservedCount {
+		panic("invalid reservedNo exceed ReservedCount")
+	}
+	return math.MaxUint64 - reservedNo
+}
+
+// GlobalAllocator is global connection ID allocator.
+type GlobalAllocator struct {
+	is64bits       sync2.AtomicInt32 // !0: true, 0: false
+	serverIDGetter func() uint64
+
+	local32 LockFreeCircularPool
+	local64 AutoIncPool
+}
+
+// Is64 indicates allocate 64bits global connection ID or not.
+func (g *GlobalAllocator) Is64() bool {
+	return g.is64bits.Get() != 0
+}
+
+// UpgradeTo64 upgrade allocator to 64bits.
+func (g *GlobalAllocator) UpgradeTo64() {
+	g.is64bits.Set(1)
+}
+
+// LocalConnIDAllocator64TryCount is the try count of 64bits local connID allocation.
+const LocalConnIDAllocator64TryCount = 10
+
+// NewGlobalAllocator creates a GlobalAllocator.
+func NewGlobalAllocator(serverIDGetter serverIDGetterFn) *GlobalAllocator {
+	g := &GlobalAllocator{
+		serverIDGetter: serverIDGetter,
+	}
+	g.local32.InitExt(1<<LocalConnIDBits32, math.MaxUint32)
+	g.local64.InitExt((1<<LocalConnIDBits64)-ReservedCount, true, LocalConnIDAllocator64TryCount)
+
+	g.is64bits.Set(1) // TODO: set 32bits as default, after 32bits logics is fully implemented and tested.
+	return g
+}
+
+// NextID returns next connection ID.
+func (g *GlobalAllocator) NextID() uint64 {
+	globalConnID := g.Allocate()
+	return globalConnID.ToConnID()
+}
+
+// GetReservedConnID implements ConnIDAllocator interface.
+func (g *GlobalAllocator) GetReservedConnID(reservedNo uint64) uint64 {
+	if reservedNo >= ReservedCount {
+		panic("invalid reservedNo exceed ReservedCount")
+	}
+
+	serverID := g.serverIDGetter()
+	globalConnID := GCID{
+		ServerID:    serverID,
+		LocalConnID: (1 << LocalConnIDBits64) - 1 - reservedNo,
+		Is64bits:    true,
```

Comment:
We always use 64bits for the reserved conn ID. Will that bring any problem?

Reply:
No.

A 64 bits global connection ID with server ID less than `MaxServerID32` (`2^11 - 1`) is still valid, and globally unique.

This scenario will also happen after upgrade from 32 bits to 64  bits when the 20 bits local ID pool has been used up, while server ID stay unchanged.